modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/client/blockstream/BlockStreamSubscription.kt

```diff
@@ -0,0 +1,137 @@
+package jp.co.soramitsu.iroha2.client.blockstream
+
+import io.ktor.client.plugins.websocket.webSocket
+import io.ktor.websocket.close
+import io.ktor.websocket.readBytes
+import jp.co.soramitsu.iroha2.IrohaSdkException
+import jp.co.soramitsu.iroha2.SingletonHolder
+import jp.co.soramitsu.iroha2.cast
+import jp.co.soramitsu.iroha2.client.Iroha2Client
+import jp.co.soramitsu.iroha2.generated.BlockSubscriptionRequest
+import jp.co.soramitsu.iroha2.generated.VersionedBlockMessage
+import jp.co.soramitsu.iroha2.generated.VersionedBlockSubscriptionRequest
+import jp.co.soramitsu.iroha2.toFrame
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.FlowCollector
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+import org.slf4j.LoggerFactory
+import java.math.BigInteger
+import java.util.UUID
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.coroutines.CoroutineContext
+
+class BlockStreamSubscription private constructor(private val context: BlockStreamContext) : CoroutineScope {
+
+    override val coroutineContext: CoroutineContext = Dispatchers.IO
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    private val source: ConcurrentHashMap<UUID, BlockStreamStorage> = ConcurrentHashMap()
+    private var stopped: AtomicBoolean = AtomicBoolean(false)
+    private var initialized: AtomicBoolean = AtomicBoolean(false)
+
+    init {
+        run()
+    }
+
+    fun subscribe(): Pair<UUID?, BlockStreamSubscription> {
+        return when (initialized.get()) {
+            false -> {
+                initialized.set(true)
+                subscribe(
+                    context.storage.closeIf,
+                    context.storage.onFailure,
+                    context.storage.onBlock,
+                ) to getInstance(context)
+            }
+            true -> null to getInstance(context)
+        }
+    }
+
+    @JvmOverloads
+    fun <T> subscribeAndReceive(
+        closeIf: (suspend (block: VersionedBlockMessage) -> Boolean)? = null,
+        onFailure: (suspend (t: Throwable) -> Unit)? = null,
+        onBlock: (block: VersionedBlockMessage) -> Any,
+        collector: FlowCollector<T>,
```

Comment:
I guess we can move collector to an optional part of the context?

Reply:
collector isn't a part of the storage

Reply:
not clear why to me

Reply:
it's just not presented in BlockStreamStorage